astro/transform.py

```diff
@@ -231,4 +231,90 @@ def cartesian_to_classical(self, r, v):
             w = float('nan')
             TA = float('nan')
 
-        return a, e, i, node, w, TA
\ No newline at end of file
+        return a, e, i, node, w, TA
+
+    def ra_dec_to_cartesian(self, r_mag, ra, dec, dr_mag, dra, ddec):
+        """ Convert right ascension and declination to cartesian elements
+
+        Args:
+            r_mag (float) - radius of object
+            ra (float) - right ascension (rad)
+            dec (float) - declination (rad)
+            dr_mag (float) - radius rate of object
+            dra (float) - right ascension rate (rad/s)
+            ddec (float) - right declination rate (rad/s)
+
+        Returns:
+            r (numpy array) - cartesian position
+            v (numpy array) - cartesian velocity
+        """
+
+        # Get position vector
+        r = np.array([0.0] * 3)
+        r[0] = r_mag * cos(dec) * cos(ra)
+        r[1] = r_mag * cos(dec) * sin(ra)
+        r[2] = r_mag * sin(dec)
+
+        # Get velocity vector
+        v = np.array([0.0] * 3)
+        v[0] = dr_mag * cos(dec) * cos(ra) \
+               - r_mag * sin(dec) * cos(ra) * ddec \
+               - r_mag * cos(dec) * sin(ra) * dra
+        v[1] = dr_mag * cos(dec) * sin(ra) \
+               - r_mag * sin(dec) * sin(ra) * ddec \
+               + r_mag * cos(dec) * cos(ra) * dra
+        v[2] = dr_mag * sin(dec) + r_mag * cos(dec) * ddec
+
+        return r, v
+
+    def cartesian_to_ra_dec(self, r, v):
+        """ Convert cartesian elements to right ascension and declination
+
+        Note that r, v do not have to be in km but need to be consistent
+
+        Args:
+            r (list or array) - cartesian position (km)
+            v (list or array) - cartesian velocity (km/s)
+
+        Returns:
+            r_mag (float) - radius of object
+            ra (float) - right ascension (rad)
+            dec (float) - declination (rad)
+            dr_mag (float) - radius rate of object
+            dra (float) - right ascension rate (rad/s)
+            ddec (float) - right declination rate (rad/s)
+        """
+
+        # Enforce numpy arrays
+        r = np.array(r)
+        v = np.array(v)
+
+        r_mag = np.linalg.norm(r)
+        temp1 = sqrt(r[0] ** 2 + r[1] ** 2)
+
+        # Get right ascension
+        if temp1 < TOL:
+            ra = atan2(v[1], v[0])
```

Comment:
I hadn't seen using velocity if the value is too small.. that's cool. Is there a reference? (Maybe we should just put references in the code?)

I confirmed that in python v3 atan2 is (y,x).... in MS Excel, I think it's (x,y!)
https://docs.python.org/3/library/math.html

Reply:
Yeah I put it at the top of the script that the reference is Vallado  😃